File: 4/p4.py
#!/usr/bin/python

import gc
import math

def pals() :
	palindromes = list()
	gc.disable() #it can potentially slow down the loop interation
	for a in range(1,10) :
		for b in range(10):
			for c in range (10):
				number = a*(1+ 10**5) + b*(10**4+10) + c*(10**2 + 10**3)
				palindromes.append(number)
	gc.enable()
	return palindromes


def FermatFactorization(N) :
	a = math.ceil(math.sqrt(N))
	b = a**2 -N

	while not (int(math.sqrt(b)) == math.sqrt(b)) : # test that B is a square, could probably be done more efficiently
		a = a+ 1
		b = a**2 - N 
	return a-math.sqrt(b),a+math.sqrt(b)


palindromes = [p for p in reversed(pals())]
for i in palindromes :
	print 'Running for '+str(i)
	res = FermatFactorization(i)
	if len(str(int(res[0])))  == 3  and len(str(int(res[1]))) == 3 :
		print 'First palindrome which is a product of two 3-digit numbers is '+str(i)
		print str(i)+' : '+str(res[0])+', '+str(res[1])
		break

File: 4/p4.c
#include <stdio.h>
#include <math.h>
#include <p4.h>

int main(void){
	
	factors ff ;

	int palindromes[899];
	generatePalindromes(palindromes);
	int b ;

	for (b = 899 ; b >= 0 ; b--){

		printf("Running for %d...\n", palindromes[b]);
		FermatFact(&ff , palindromes[b]);
		
		if( ff.a > 100  && ff.a < 1000 && ff.b > 100 &&ff.b < 1000 ){
			printf("    Found palindrome with 3-digit factors, it is %d, factors (%d,%d) \n", palindromes[b], ff.a,ff.b);
			break;
		}
	}

	return 0 ;
}

void generatePalindromes( int *arr){
	int a ; 
	int b ;
	int c;
	for (a = 9 ; a > 0 ; a--){
		for (b = 9 ; b >= 0 ; b--){
			for (c = 9 ; c >= 0 ;c--){
				arr[a*100+b*10+c - 100] = a*(1+(int) pow(10,5))+b*(10+(int) pow(10,4))+c*((int) pow(10,2) + (int) pow(10,3));
			}
		}
	}
}

void FermatFact(factors* f , int N){
	double a = ceil(sqrt(N));
	double b = a*a-N ;

	while (sqrt(b) != ceil(sqrt(b)) ){
		a = a + 1 ;
		b = a*a -N ;
	}

	f->a = (int) a-sqrt(b) ; 
	f->b = (int) a+sqrt(b) ; 
}
